src/org/mozilla/javascript/NativeJSON.java

```diff
@@ -500,6 +550,28 @@ private static String quote(String string) {
         return product.toString();
     }
 
+    private static Object javaToJSON(Object value, StringifyState state) {
+        value = state.cx.getJavaToJSONConverter().apply(value);
+        value = Context.javaToJS(value, state.scope, state.cx);
```

Comment:
I still think about the jackson use case. 

What do you think is the best method to use jackson?

1. There is a `ObjectMapper.convertValue(bean, Map.class)` https://fasterxml.github.io/jackson-databind/javadoc/2.8/com/fasterxml/jackson/databind/ObjectMapper.html#convertValue(java.lang.Object,%20java.lang.Class) which seems to be useful for this task but may produce a lot of Map/List instances.

2. Using writeValueAsString() will be more efficient, but the returned value will be currently treated as normal string and gets quoted again, so should we add the ability to return a raw json?

This could be done by adding a simple transport-class
```java
public class RawJson {
   private String json;
   public RawJson(String json) { this.json = json }
}
```
the converter then would look like `UnaryOperator<Object> JACKSON v -> new RawJson(objectMapper.writeValueAsString(v))` and here we add an "instanceOf" check, if it is already a raw json.

Pro: this will be probably more performant and do not need to instanttiagea huge amount of List/Maps during conversions

Con: you loose the ability to use replacer or indention options of JSON.stringify. 

On the other hand you may call `objectMapper.writeValueAsString` also from the javascript side (but jackson may not be able to deserialize JavaScript objects)

Reply:
`ObjectMapper.convertValue(bean, Map.class)` is what I had in mind.

I had thought about passing the state along with the object to the converter, but then decided against it to keep it simple. By having the return value stringified, that allowed for:
- js implementations to return js objects
- java implementations to return Maps/Collections/Arrays for objects containing other objects
- stringify handles recursion instead of the converter
- Scriptables contained in java objects to be stringified (or not for functions, symbols, etc..) as the correct js type
- returning Undefined.instance to explicitly drop the property containing this object from the final output
- respecting the current stringify state
- strings to be automatically quoted and escaped

Would it be worth seeing if the current behavior is a performance issue before introducing RawJson in another PR?

Only introducing RawJson I think would be tolerable with its shortcomings since it would not be used by the default implementation, and would only add an additional type check. An implementation could also output something that isn't actually valid JSON with this option.

Getting it to play nice with the replacer and indent would be more difficult.
- The StringifyState would need to be passed to the converter. It should probably be a clone of the actual state so that the converter doesn't corrupt it. This would also lead to the creation of a lot of objects, even though the provided implementations would not use them.
-  StringifyState would need to be made public and shouldn't be an inner class.
-  The converter type would need to be changed from a UnaryOperator to a BiFunction to support passing the additional argument.
- The converter itself would need to implement handling of the replacer and pretty printing.

Reply:
StringifyState could implement an interface which exposes only getters. This should avoid corruption (if no reflection is used) and no new object has to be created.

Reply:
StringifyState has properties that are not immutable, though. I think I'd rather not go down that path, as it complicates the implementation to get a potential performance boost for an edge case, which isn't even possible right now (stringifying a high volume of java objects using a custom implementation.)

Is it possible to attack this from the other end and teach Jackson how to handle Scriptables?

Reply:
My idea was to implement immutable getters, e.g. `StringifyState.getPropertyList()` could return `Collections.unmodifiableList(...)`.

But all in all, I agree, that this edge case would add lot of complexity.
Currently, I see the following options:

- Use JSON.stringify with Jackson's convertValue. This will work in all cases and will support all stringify-options (like replacer, indention options...) but adds some overhead.
- Use Jackson directly. This gives you the best performance. You may have to write special JsonSerializer for some JS-objects. (NativeObject/NativeArray should already work out of the box, because they implement Map/List)
- You can use RawJson. (if implemented)

> Would it be worth seeing if the current behavior is a performance issue before introducing RawJson in another PR?
>
> Only introducing RawJson I think would be tolerable with its shortcomings since it would not be used by the default implementation, and would only add an additional type check. An implementation could also output something that isn't actually valid JSON with this option.

If you have time, you can do some measurements, what the impact of convertValue vs writeValueAsString is.
If we really see, that it might be a performance issue, we can implement RawJson (without passing the state) and document the shortcomes. But this can be done also in a later PR.

From my point of view everything is discussed and considered. I would accept this PR now.

Reply:
FWIW Rhino is used in so many places, and version management has become so complex that depending on Jacskon in the core of Rhino would be hard for a lot of people to adopt.

However, if we created another package from this project, like, say "rhino-jackson-adapter", and by putting that JAR in the classpath it causes Rhino to use Jackson instead for serialization, then people who want better performance and can change their build rules can just use that. I have always been a fan of Java's "ServiceLoader" pattern for things like this but there might be better ways today.

The conversation about string concatenation is also fascinating -- I have been so indoctrinated to never concatenate strings in loops using + because of all the garbage that gets created!